### PR TITLE
Added option to create alert filters from alert

### PR DESCRIPTION
**Pull request by psiinon**
Also added the option to test and apply alert filters.
Still WIP as I need to update the help, but I think the code is ready to review.

Fixes https://github.com/zaproxy/zaproxy/issues/3027 (at least I think it does;)

Signed-off-by: Simon Bennetts <psiinon@gmail.com>

### CODE REVIEW

**Comment by psiinon** [comment]
Could we promote this add-on to Release when this PR is merged?

**Comment by kingthorin** [comment]
I support that idea.

**Comment by kingthorin** [comment]
I think you can also mark this as fixing https://github.com/zaproxy/zaproxy/issues/2599 because it allows to modify existing alerts? (Assuming I'm understanding correctly...I haven't read through the code yet.)

**Comment by psiinon** [comment]
All planned changes now made, so ready for full review.
@kingthorin I'll double check https://github.com/zaproxy/zaproxy/issues/2599 and see if this fixes it...

**Comment by psiinon** [comment]
Re https://github.com/zaproxy/zaproxy/issues/2599 .. I'm not really sure ;) We could ask the person who raised it once this is merged...

**Comment by psiinon** [comment]
Ugh, just done some more testing and the global alerts are not saving correctly :(

**Comment by psiinon** [comment]
Just a bug I introduced in my last set of changes - fixed now :)

File: addOns/alertFilters/src/main/java/org/zaproxy/zap/extension/alertFilters/AlertFiltersMultipleOptionsPanel.java [hunk 1]
```diff
@@ -46,12 +46,15 @@
...
             Constant.messages.getString("alertFilters.dialog.remove.checkbox.label");
 
+    private ExtensionAlertFilters extension;
     private DialogAddAlertFilter addDialog = null;
     private DialogModifyAlertFilter modifyDialog = null;
     private Context uiSharedContext;
 
-    public AlertFiltersMultipleOptionsPanel(AlertFilterTableModel model) {
+    public AlertFiltersMultipleOptionsPanel(
+            ExtensionAlertFilters extension, AlertFilterTableModel model) {
```
Comment (kingthorin) [inline addOns/alertFilters/src/main/java/org/zaproxy/zap/extension/alertFilters/AlertFiltersMultipleOptionsPanel.java]: Are we worried about changing this signature?

Do we really need to pass the extension around for code that is in the extension?

**Comment by thc202** [comment]
Thanks!

File: addOns/alertFilters/src/main/resources/org/zaproxy/zap/extension/alertFilters/resources/Messages.properties [hunk 2]
```diff
@@ -24,6 +25,17 @@ alertFilters.dialog.add.field.label.attack		= Attack:
...
+alertFilters.dialog.add.field.label.apply		= Apply Filter:
+
+alertFilters.dialog.add.label.scope.global		= Global
+
+alertFilters.dialog.button.apply	= Apply
+alertFilters.dialog.button.test		= Test
+alertFilters.dialog.filter.state.notapplied = Not applied
+alertFilters.dialog.filter.state.nottested = Not tested
+alertFilters.dialog.filter.state.appliesto = Applies to {0} alerts
+alertFilters.dialog.filter.state.appliedto = Applied to {0} alerts
```
Comment (kingthorin) [inline addOns/alertFilters/src/main/resources/org/zaproxy/zap/extension/alertFilters/resources/Messages.properties]: Duplicate

Comment (psiinon) [reply addOns/alertFilters/src/main/resources/org/zaproxy/zap/extension/alertFilters/resources/Messages.properties]: Its not actually - "Applies" vs "Applied" :)

File: addOns/alertFilters/src/main/java/org/zaproxy/zap/extension/alertFilters/AlertFiltersMultipleOptionsPanel.java (hunk 1 above)
Comment (psiinon) [reply addOns/alertFilters/src/main/java/org/zaproxy/zap/extension/alertFilters/AlertFiltersMultipleOptionsPanel.java]: I'm not, and we need a reference to it somehow. Pretty its done like that in a load of other extensions ;)

File: addOns/alertFilters/src/main/resources/org/zaproxy/zap/extension/alertFilters/resources/Messages.properties (hunk 2 above)
Comment (kingthorin) [reply addOns/alertFilters/src/main/resources/org/zaproxy/zap/extension/alertFilters/resources/Messages.properties]: OMG! Yikes... Okay

**Comment by psiinon** [comment]
@kingthorin any other changes needed?

**Comment by kingthorin** [comment]
Nope, LGTM

**Comment by kingthorin** [comment]
I could know the release off this morning if we’re ready for that?

**Comment by psiinon** [comment]
So upgrade to Release quality and release? I'm good with that :)

**Comment by kingthorin** [comment]
On it.

**Pull request merged**